### PR TITLE
(feat) User list fetch api

### PR DESCRIPTION
**Pull request by shulhasupertokens**
## Summary

- Display users into table by fetching data from Users and Count API
- Implement pagination by utilizing `goToNext` & `offsetChange` prop
- `goToNext` will trigger the API call to load the next page and store it in local memory
-  `offsetChange`  will trigger loading records from local memory based on the offset
- Fix element`#root`'s  background color
- Create loading indicator using placeholder as it will also minimize layout shifting
- Apply SCSS for better CSS structure

### CODE REVIEW

File: src/ui/components/usersListTable/UsersListTable.tsx [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+  const { limit, offset, goToNext, offsetChange, nextPaginationToken, isLoading, count } = {
+    offset: 0,
+    limit: LIST_DEFAULT_LIMIT,
+    ...props,
+  }
+  return (
+    <div className='users-list-pagination'>
+      <p className='users-list-pagination-count text-small'>
+        {offset + 1}- {offset + limit} of {count}
+      </p>
```
Comment (nkshah2) [inline src/ui/components/usersListTable/UsersListTable.tsx:129]: Try to add comments to your code (in this and other places)

File: src/ui/pages/usersList/UsersList.css [hunk 2]
```diff
@@ -0,0 +1,24 @@
+.users-list {
+  margin: 0 auto;
+  padding: 30px 0;
+  width: calc(100% - 32px);
+  max-width: 829px;
```
Comment (nkshah2) [inline src/ui/pages/usersList/UsersList.css:5]: Where did this number come from?

File: src/ui/pages/usersList/types.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+export enum UserRecipeType { 
+  emailpassword= 'emailpassword', 
+  thirdparty = "thirdparty", 
+  passwordless = "passwordless" }
+
+export type User = { recipeId: UserRecipeType.emailpassword, user: UserEmailPassword } | 
+  { recipeId: UserRecipeType.passwordless, user: UserPasswordless } |
+  { recipeId: UserRecipeType.thirdparty, user: UserThirdParty }
+
+export interface UserEmailPassword {
```
Comment (nkshah2) [inline src/ui/pages/usersList/types.ts:10]: Why do we need an interface for this?

File: src/ui/pages/usersList/types.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+  email?: string
+  phoneNumber?: string
+  timeJoined?: number
+  firstName?: string
+  lastName?: string
+}
+
+export interface UserPasswordless extends UserEmailPassword {}
+
+export interface UserThirdParty extends UserEmailPassword {
```
Comment (nkshah2) [inline src/ui/pages/usersList/types.ts:21]: Why not just use types for this?

File: src/ui/utils/number.ts [hunk 5]
```diff
@@ -0,0 +1,5 @@
+export const ordinal = (num: number) => {
+  const mod = num % 10
+  const modMap: Record<number, string> = {1: 'st',2: 'nd',3: 'rd'}
+  return num >10 && num < 14 ? 'th' : (modMap[mod] ?? 'th')
+}
```
Comment (nkshah2) [inline src/ui/utils/number.ts:5]: We dont need separate files like this, just organise them in a single utils file

**Review by nkshah2 — CHANGES_REQUESTED** [review]

File: src/ui/utils/number.ts (hunk 5 above)
Comment (shulhasupertokens) [reply src/ui/utils/number.ts:5]: ordinal is not Date specific utils. Also, in the future, this file can be the container for another helper that is related to number

File: src/ui/pages/usersList/UsersList.css (hunk 2 above)
Comment (shulhasupertokens) [reply src/ui/pages/usersList/UsersList.css:5]: image
The Zeplin says its 829px

File: src/App.tsx [hunk 6]
```diff
@@ -3,43 +3,24 @@ import {
...
   return (
     <Router basename={getDashboardAppBasePath()}>
       <Routes>
         <Route path='/auth' element={<Auth />}/>
-        <Route path='/' element={<Home/>} />
+        <Route path='/' element={(
+          <AuthWrapper>
+            <UsersList />
+          </AuthWrapper>
+        )} />
```
[the remark this answers is not in the document]
Comment (shulhasupertokens) [reply src/App.tsx]: Sorry, do you mean to move this `AuthWrapper` into `UserList` container component ?

File: src/ui/components/phoneNumber/PhoneNumber.scss [hunk 7]
```diff
@@ -0,0 +1,9 @@
+.phone-display {
+  display: flex;
+  align-items: center;
+  gap: 5px;
```
[the remark this answers is not in the document]
Comment (shulhasupertokens) [reply src/ui/components/phoneNumber/PhoneNumber.scss:4]: Okay, created `gap-ios` mixin to fix this 

reference https://ppuzio.medium.com/flexbox-gap-workaround-for-safari-on-ios-14-13-and-lower-ffcae589eb69

File: src/ui/pages/usersList/types.ts (hunk 3 above)
Comment (shulhasupertokens) [reply src/ui/pages/usersList/types.ts:10]: This to make the  `User` Type to determine automatically the correct user properties based on the `recipeId`.

Is this answering your questions?

File: api_spec.yaml [hunk 8]
```diff
@@ -115,7 +115,7 @@ paths:
                                 email:
                                   type: string
                                   example: "johndoe@gmail.com"
-                                phone:
+                                phoneNumber:
```
Comment (shulhasupertokens) [inline api_spec.yaml:118]: This changes to reflect the correct property naming based on the returned data from API

File: src/ui/components/usersListTable/UsersListTable.tsx (hunk 1 above)
Comment (shulhasupertokens) [reply src/ui/components/usersListTable/UsersListTable.tsx]: I have break it down to smaller functions

File: src/App.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (shulhasupertokens) [reply src/App.tsx]: Its moved

File: api_spec.yaml [hunk 9]
```diff
@@ -100,7 +100,7 @@ paths:
                                   type: string
                                 lastName:
                                   type: string
-                                thirdparty:
+                                thirdParty:
```
Comment (shulhasupertokens) [inline api_spec.yaml:103]: This changes to reflect the correct property naming based on the returned data from API

File: src/ui/components/phoneNumber/PhoneNumber.scss [hunk 10]
```diff
@@ -0,0 +1,13 @@
+@import "../../styles/mixin.scss";
+
+.phone-display {
+  display: flex;
+  align-items: center;
+  gap: 5px;
+  @include gap-ios(5px);
```
Comment (nkshah2) [inline src/ui/components/phoneNumber/PhoneNumber.scss:7]: I would prefer if we just dont use gap instead. Doing it this way just adds more boilerplate and theres a high chance that someone unfamiliar with this project will just miss out on adding the mixin

File: src/ui/components/usersListTable/UsersListTable.tsx [hunk 11]
```diff
@@ -1,63 +1,181 @@
...
-            <td>Date and time of joining</td>
-          </tr>
-          <tr>
-            <td>John Doe</td>
-            <td>Auth Method Pill</td>
-            <td>Date and time of joining</td>
-          </tr>
+        <tbody className='text-small'>
+          {
+            isLoading && PlaceholderTableRows() // show placeholder when it is loading from API
```
Comment (nkshah2) [inline src/ui/components/usersListTable/UsersListTable.tsx:42]: Why not just use it as a component?

File: src/ui/pages/usersList/UsersList.css (hunk 2 above)
Comment (nkshah2) [reply src/ui/pages/usersList/UsersList.css:5]: Its a bad idea to use numbers directly from Zeplin like this, thats just the width of the frame used in the designs. Just use a percentage of the screen width instead

File: src/ui/pages/usersList/types.ts (hunk 3 above)
Comment (nkshah2) [reply src/ui/pages/usersList/types.ts:10]: No I meant why use an interface and not a type?

File: src/ui/pages/usersList/types.ts (hunk 4 above)
Comment (nkshah2) [reply src/ui/pages/usersList/types.ts:19]: Also its strange to see Passwordless inherit from email password, ideally you want

```ts
type User = {...}

type EmailPasswordUser = User & {...}

type PasswordlessUser = User & {...}
```

and so on. You can use type unions and exclusions for addition/removal of properties

File: src/ui/utils/number.ts (hunk 5 above)
Comment (nkshah2) [reply src/ui/utils/number.ts]: Id much rather have all utils in the same file, if it helps just use a class with static methods to segregate within that file. If a specific type has several helpers in the future, we can consider moving that to its own file

Comment (shulhasupertokens) [reply src/ui/utils/number.ts]: Its merged in src\utils.ts

File: src/ui/pages/usersList/types.ts (hunk 4 above)
Comment (shulhasupertokens) [reply src/ui/pages/usersList/types.ts]: Its Updated

File: src/ui/pages/usersList/UsersList.css (hunk 2 above)
Comment (shulhasupertokens) [reply src/ui/pages/usersList/UsersList.css]: added rule width:90% if screen's width > 900px

File: src/ui/components/phoneNumber/PhoneNumber.scss (hunk 7 above)
[the remark this answers is not in the document]
Comment (shulhasupertokens) [reply src/ui/components/phoneNumber/PhoneNumber.scss]: the gap CSS is removed

File: src/types.ts [hunk 12]
```diff
@@ -13,4 +13,42 @@
...
-export type HttpMethod = "GET" | "POST";
\ No newline at end of file
+// HTTP requests & responses related 
+export type HttpMethod = "GET" | "POST";
+export interface ListCount {count: number}
+export interface ResponseList<T>  {
+    nextPaginationToken?: string
+    status: Response['status'],
+    users: T[]
+}
```
Comment (shulhasupertokens) [inline src/types.ts:23]: @nkshah2 In next phase, I think we should format the API's list response to be like this(replace `users` field to be `records`):
```
export interface ResponseList<T>  {
    nextPaginationToken?: string
    status: Response['status'],
    records: T[]
}
```

So, that later we can use a generic `ResponseList` for different kind of list.

File: src/types.ts [hunk 13]
```diff
@@ -13,4 +13,42 @@
...
+  lastName?: string
+}
+
+export type UserEmailPassword = UserProperties
+
+export type UserPasswordLess = UserProperties
+
+export type UserThirdParty = UserProperties & {
+  thirdParty: {
+    id: 'google' | 'github' | string
```
Comment (nkshah2) [inline src/types.ts:51]: Missing apple and facebook (https://github.com/supertokens/supertokens-node/tree/master/lib/ts/recipe/thirdparty/providers)

File: src/types.ts [hunk 14]
```diff
@@ -13,4 +13,42 @@
...
+
+export type User = { recipeId: UserRecipeType.emailpassword, user: UserEmailPassword } | 
+  { recipeId: UserRecipeType.passwordless, user: UserPasswordLess } |
+  { recipeId: UserRecipeType.thirdparty, user: UserThirdParty }
+
+export interface UserProperties {
+  id: string
+  email?: string
+  phoneNumber?: string
+  timeJoined?: number
```
Comment (nkshah2) [inline src/types.ts:40]: Shouldnt be optional

File: src/types.ts [hunk 15]
```diff
@@ -13,4 +13,42 @@
...
+export enum UserRecipeType { 
+  emailpassword= 'emailpassword', 
+  thirdparty = "thirdparty", 
+  passwordless = "passwordless" }
+
+export type User = { recipeId: UserRecipeType.emailpassword, user: UserEmailPassword } | 
+  { recipeId: UserRecipeType.passwordless, user: UserPasswordLess } |
+  { recipeId: UserRecipeType.thirdparty, user: UserThirdParty }
+
+export interface UserProperties {
```
Comment (nkshah2) [inline src/types.ts:36]: Rename to `User` or `UserType`

Comment (nkshah2) [inline src/types.ts:36]: Change all interfaces to use types

File: src/types.ts [hunk 16]
```diff
@@ -13,4 +13,42 @@
...
+export interface ListCount {count: number}
+export interface ResponseList<T>  {
+    nextPaginationToken?: string
+    status: Response['status'],
+    users: T[]
+}
+
+
+// Users Models
+export enum UserRecipeType {
```
Comment (nkshah2) [inline src/types.ts:27]: Use this format instead

```ts
export type UserRecipeType = "emailpassword" | "..."
```

File: src/types.ts (hunk 12 above)
Comment (nkshah2) [reply src/types.ts:23]: We dont need to do this

File: src/ui/pages/usersList/UsersList.tsx [hunk 17]
```diff
@@ -1,23 +1,92 @@
...
+  )
 }
 
-export default UsersList;
+const fetchUsers = async (param?: { paginationToken?: string }) =>
+  fetchDataAndRedirectIf401({
+    url: getApiUrl('/api/users'),
+    method: 'GET',
+    query: { limit: `${LIST_DEFAULT_LIMIT}`, ...param },
+  }).then(async (response) => response && ((await response?.json()) as ResponseList<User>))
```
Comment (nkshah2) [inline src/ui/pages/usersList/UsersList.tsx:78]: Use async await

File: src/types.ts [hunk 18]
```diff
@@ -13,4 +13,42 @@
 * under the License.
 */
 
-export type HttpMethod = "GET" | "POST";
\ No newline at end of file
+// HTTP requests & responses related 
+export type HttpMethod = "GET" | "POST";
+export interface ListCount {count: number}
```
Comment (nkshah2) [inline src/types.ts:18]: Rename types to be more specific, and move them to a types file in the page/component

File: src/types.ts [hunk 19]
```diff
@@ -13,4 +13,42 @@
...
+  passwordless = "passwordless" }
+
+export type User = { recipeId: UserRecipeType.emailpassword, user: UserEmailPassword } | 
+  { recipeId: UserRecipeType.passwordless, user: UserPasswordLess } |
+  { recipeId: UserRecipeType.thirdparty, user: UserThirdParty }
+
+export interface UserProperties {
+  id: string
+  email?: string
+  phoneNumber?: string
```
Comment (nkshah2) [inline src/types.ts:39]: EmailPassword and ThirdParty dont have phoneNumber

**Review by nkshah2 — CHANGES_REQUESTED** [review]

File: src/types.ts (hunk 18 above)
Comment (shulhasupertokens) [reply src/types.ts:18]: ok

File: src/types.ts (hunk 12 above)
Comment (shulhasupertokens) [reply src/types.ts:23]: ok

**Review by nkshah2 — APPROVED** [review]

**Pull request merged**